API definition.

mroman-redhat opened the pull request:
Added a swagger API definition for UI purposes.

A couple of notes:
- Username will be set in the UI not the API
- Local/static file requests are handled primitively, however I assumed its not necessary to rework, since the API will only ever receive requests for 3 known files.
- the `openshift` folder contains files needed to run the api on the cluster
- The API has the following capabilities:
   - Get/Update a user ConfigMap (GPU, and EnvVar reading is included in this, will split if necessary)
   - Get Sizes (Both full json readout of just the names), Also get a single size
   - Get Images (Names only)

Information for testing is in the API section of Readme

mroman-redhat commented:
I just realised it was neccessary to "hack" the service.py file, what should be done about that?

vpavlin commented on openshift/api-service.yaml:
```diff
@@ -0,0 +1,16 @@
+      targetPort: 8080
+  selector:
+    app: api-app
+  clusterIP: 172.30.63.15
```
ClusterIP should not be in a general service manifest

vpavlin commented on openshift/api-build.yaml:
```diff
@@ -0,0 +1,30 @@
+    type: Git
+    git:
+      uri: 'https://github.com/mroman-redhat/jupyterhub-singleuser-profiles.git'
+      ref: Issue-13
```
SHould it be 30?

vpavlin commented:
Can you add a route to the openshift directory as well?

It is also missing serviceaccount, role and rolebinding to be able to access and modify ConfigMaps

vpavlin commented:
It would be also good to add example profiles ConfigMap so that someone trying this out does not get "ConfigMap not found error"

mroman-redhat commented:
Removed the ClusterIP, added Route, Service Account, custom Role and Rolebinding, and an example ConfigMap for testing. The API should be easier to test now.

vpavlin commented on jupyterhub_singleuser_profiles/images.py:
```diff
@@ -16,7 +17,7 @@ def get_images(self, label=None):
         imagestreams = self.oapi_client.resources.get(kind='ImageStream', api_version='image.openshift.io/v1')
         imagestream_list = imagestreams.get(namespace=self.namespace, label_selector=label)
-
+        
```
Remove whitespaces

vpavlin commented on app.py:
```diff
@@ -0,0 +1,55 @@
+
+app = connexion.App(__name__, specification_dir='.', server='tornado', options={'swagger_ui':True})
+app.add_api('swagger.yaml')
+app.run(port=8080)
```
Missing new line

vpavlin commented on README.md:
```diff
@@ -196,6 +196,23 @@ You can list images in your cluster by running
+  - Sizes (`GET`), Sizes can be requested, either as an array of names (used in the UI) or as a pure json body (using the `pure_json=true` query)
+  - Images (`GET`), Returns an array of images (which have the proper tag). Note: Due to some limitation it is not possible to gather a json with all of the image information
+
+The API also has some primitive local file gathering capability for build purposes.
```
What does this mean? What capability?

mroman-redhat replied:
I wanted to mention it because I use the API to access some of the static files (the chunk.js files) and I thought it would be good to mention it if something similar was needed in the future. I never really inteded to use it beyond just accessing the index.html to actually show the page but it could be useful.

vpavlin replied:
What you described does not sound to me like what that sentence says:) I would just remove it - it is not extremely important for the actual API

vpavlin commented on README.md:
```diff
@@ -196,6 +196,23 @@ You can list images in your cluster by running
+The API also has some primitive local file gathering capability for build purposes.
+
+For testing:
+  - To deploy on your cluster use: `oc apply -f openshift`
```
Move the command to a separate line/block (use 3 ticks to get code block)

vpavlin commented on README.md:
```diff
@@ -196,6 +196,23 @@ You can list images in your cluster by running
+
+For testing:
+  - To deploy on your cluster use: `oc apply -f openshift`
+  - Wait for `api-build` to finish (or start a new one)
```
Suggest using

```
oc get builds
```

vpavlin commented on README.md:
```diff
@@ -196,6 +196,23 @@ You can list images in your cluster by running
+For testing:
+  - To deploy on your cluster use: `oc apply -f openshift`
+  - Wait for `api-build` to finish (or start a new one)
+  - If `api-deploy` is running you can access the api through the route `api-jh-route` and adding /ui to the url
```
fix the route name

Suggest using 

```
oc get route jh-api-route -o jsonpath='http://{.spec.host}/ui'
```

vpavlin commented on app.py:
```diff
@@ -0,0 +1,55 @@
+from jupyterhub_singleuser_profiles.profiles import SingleuserProfiles
+
+_PROFILES = SingleuserProfiles(verify_ssl=False)
+_PROFILES.load_profiles()
```
I believe we said we need to run load_profiles before each request to make sure things are up to date - correct?

Maybe check this: https://github.com/zalando/connexion/issues/906

vpavlin commented on jupyterhub_singleuser_profiles/images.py:
```diff
@@ -37,6 +35,22 @@ def append_option(self, image, result, last_image):
                 selected = "selected=selected"
             result.append("<option value='%s' %s>%s</option>" % (image_tag, selected, image_tag))
 
+    def get_list(self):
```
Can you refactor this and append_option so that the same code is there only once?

vpavlin commented on jupyterhub_singleuser_profiles/profiles.py:
```diff
@@ -120,7 +119,7 @@ def load_profiles(self, secret_name="jupyter-singleuser-profiles", filename=None
           self.profiles.extend(config_map_yaml.get("profiles", [self.empty_profile()]))
         else:
-          _LOGGER.error("Could not find config map %s" % config_map_name)
+          _LOGGER.error("Could not find config map")
```
can you jest fix the variable name? It should be `cm_name`

vpavlin commented on openshift/api-build.yaml:
```diff
@@ -0,0 +1,30 @@
+  source:
+    type: Git
+    git:
+      uri: 'https://github.com/mroman-redhat/jupyterhub-singleuser-profiles.git'
```
This should point to the upstream url and the ref should be master

mroman-redhat replied:
Planned to change this right before merging, otherwise the tests would not work.

vpavlin commented on jupyterhub_singleuser_profiles/images.py:
```diff
@@ -47,7 +47,10 @@ def get_form(self, last_image=None):
         else:
             for i in imagestream_list.items:
-                self.append_option(i, result, last_image)
+                if name_only:
```
Do you need this condition? Can you just pass in the name_only and let the append_option check True/False (as it does anyway)

vpavlin commented on jupyterhub_singleuser_profiles/profiles.py:
```diff
@@ -120,7 +119,7 @@ def load_profiles(self, secret_name="jupyter-singleuser-profiles", filename=None
           self.profiles.extend(config_map_yaml.get("profiles", [self.empty_profile()]))
         else:
-          _LOGGER.error("Could not find config map %s" % config_map_name)
+          _LOGGER.error("Could not find config map %s", cm_name)
```
It should be `%` not `,`
